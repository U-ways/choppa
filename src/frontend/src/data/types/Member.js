export default class Member {
  constructor(id, name, chapter) {
    this._id = id;
    this._name = name;
    this._chapter = chapter;
  }

  get id() {
    return this._id;
  }

  get name() {
    return this._name;
  }

  set name(newName) {
    this._name = newName;
  }

  get chapter() {
    return this._chapter;
  }
}
