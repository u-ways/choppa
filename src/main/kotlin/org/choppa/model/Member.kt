package org.choppa.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.relations.IterationHistory
import org.choppa.model.relations.SquadCurrentMembers
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @OneToMany(mappedBy = "member", fetch = EAGER)
    @JsonIgnore
    var squads: MutableList<SquadCurrentMembers> = mutableListOf(),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    var iterations: MutableList<IterationHistory> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
