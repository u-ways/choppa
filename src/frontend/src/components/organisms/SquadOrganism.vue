<template>
  <div class="squad">
    <div class="squad__header" :style="{ 'border-top-color': squad.colour }">
      <h1 class="squad__header-name m-0"><span class="squad__header-name--light">Squad</span> {{ squad.name }}</h1>
      <div class="row align-items-center m-0 pt-1">
        <div class="squad__header-member-count">
          {{ squad.members.length }} Members
        </div>
        <div class="squad__header-profile-pictures ml-1">
          <ProfilePictureStackMolecule :members="squad.members" :showCount="3"/>
        </div>
      </div>
    </div>
    <div class="squad__contents">
      <div class="row ml-0 mr-0 pb-1">
        <div class="col-1"></div>
        <div class="col-6">
          <span class="squad__contents--heading">Name</span>
        </div>
        <div class="col">
          <span class="squad__contents--heading">Chapter</span>
        </div>
      </div>
      <div class="row squad__contents-person-row" v-for="member in squad.members" v-bind:key="member.id">
        <div class="col-6 d-flex align-items-center squad__contents-person-cell">
          <div class="d-inline mr-3 squad__contents-person-picture">
            <ProfilePictureAtom :member="member"/>
          </div>
          <div>{{ member.name }}</div>
        </div>
        <div class="col-3 squad__contents-person-cell">
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import Squad from "@/data/types/Squad";
import ProfilePictureStackMolecule from "@/components/molecules/ProfilePictureStackMolecule";
import ProfilePictureAtom from "@/components/atoms/ProfilePictureAtom";

export default {
  name: "SquadOrganism",
  components: { ProfilePictureAtom, ProfilePictureStackMolecule },
  props: {
    squad: Squad,
  },
};
</script>

<style scoped lang="scss">
@import "../../assets/scss/colours";
@import "../../assets/scss/typography";

.squad {
  width: 100%;
  height: 100%;
  box-shadow: 0 2px 6px 0 hsla(0, 0%, 0%, 0.2);
  background-color: $light-grey;

  &__header {
    border-top: 5px solid $accent;
    background-color: $white;
    padding: 1.25rem 1.25rem 0.8rem 1.25rem;
    max-width: 100%;

    &-name {
      font-size: 1.5rem;
      font-weight: $font-weight-light;
      overflow: hidden;
      max-width: 100%;

      &--light {
        font-weight: $font-weight-extra-light;
      }
    }

    &-member-count {
      font-weight: $font-weight-extra-light;
      font-size: 0.899rem;
      display: inline-block;
    }

    &-profile-pictures {
      display: inline-block;
    }
  }

  &__contents {
    padding: 0.7rem 0 1rem 0;
    max-width: 100%;

    &--heading {
      font-weight: $font-weight-light;
      font-size: 0.75rem;
    }

    &-person-row {
      max-width: 100%;
      margin: 0;
      padding: 0.4rem 0;
      font-weight: $font-weight-medium;
      font-size: 1rem;
      color: hsla(0, 0, 0, 0.7);
      border-top: 1px solid $white-border;
    }

    &-person-cell {
      overflow: hidden;
    }

    &-person-picture {
      min-width: 2rem;
      min-height: 2rem;
      max-width: 2rem;
      max-height: 2rem;
      width: 2rem;
      height: 2rem;
    }
  }
}
</style>
