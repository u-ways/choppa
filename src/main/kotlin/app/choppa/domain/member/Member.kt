package app.choppa.domain.member

import app.choppa.domain.base.BaseModel
import app.choppa.domain.chapter.Chapter
import app.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import app.choppa.domain.squad.Squad
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @Column(name = "active", columnDefinition = "BOOLEAN")
    val active: Boolean = true,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=${chapter.name})"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
