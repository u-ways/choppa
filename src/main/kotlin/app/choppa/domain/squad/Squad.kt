package app.choppa.domain.squad

import app.choppa.domain.account.Account
import app.choppa.domain.account.Account.Companion.PLACEHOLDER_ACCOUNT
import app.choppa.domain.base.BaseModel
import app.choppa.domain.member.Member
import app.choppa.domain.member.Member.Companion.NO_MEMBERS
import app.choppa.domain.tribe.Tribe
import app.choppa.domain.tribe.Tribe.Companion.PLACEHOLDER_TRIBE
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*
import javax.persistence.FetchType.EAGER

@Entity
@Table(name = "squad")
@JsonSerialize(using = SquadSerializer::class)
@JsonDeserialize(using = SquadDeserializer::class)
data class Squad(
    @Id
    @Column(name = "squad_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String,

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int,

    // TODO(U-ways) we might be able to lazy fetch this (default is eager)
    @ManyToOne
    @JoinColumn(name = "tribe", referencedColumnName = "tribe_id")
    val tribe: Tribe,

    @ManyToMany(fetch = EAGER)
    @JoinTable(
        name = "squad_current_members",
        joinColumns = [JoinColumn(name = "squad_id", referencedColumnName = "squad_id")],
        inverseJoinColumns = [JoinColumn(name = "member_id", referencedColumnName = "member_id")]
    )
    val members: MutableList<Member> = NO_MEMBERS,

    @ManyToOne
    @JoinColumn(name = "account_id", referencedColumnName = "account_id")
    override val account: Account,
) : BaseModel {
    override fun toString() = "Squad(id=$id, name=$name, tribe=$tribe)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Squad
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_SQUADS: MutableList<Squad>
            get() = mutableListOf()
        val PLACEHOLDER_SQUAD: Squad
            get() = Squad(name = "", color = 0, tribe = PLACEHOLDER_TRIBE, account = PLACEHOLDER_ACCOUNT)
    }
}
