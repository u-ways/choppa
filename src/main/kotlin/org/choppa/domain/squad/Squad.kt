package org.choppa.domain.squad

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.history.History
import org.choppa.domain.member.Member
import org.choppa.domain.member.Member.Companion.NO_MEMBERS
import org.choppa.domain.tribe.Tribe
import org.choppa.domain.tribe.Tribe.Companion.UNASSIGNED_TRIBE
import org.choppa.utils.Color.Companion.GREY
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.fromString
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "squad")
@JsonSerialize(using = SquadSerializer::class)
@JsonDeserialize(using = SquadDeserializer::class)
data class Squad(
    @Id
    @Column(name = "squad_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "SQ-$id".substring(0, 15),

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int = GREY,

    @ManyToOne
    @JoinColumn(name = "tribe", referencedColumnName = "tribe_id")
    val tribe: Tribe = UNASSIGNED_TRIBE,

    @ManyToMany(fetch = EAGER)
    @JoinTable(
        name = "squad_current_members",
        joinColumns = [JoinColumn(name = "squad_id", referencedColumnName = "squad_id")],
        inverseJoinColumns = [JoinColumn(name = "member_id", referencedColumnName = "member_id")]
    )
    val members: MutableList<Member> = NO_MEMBERS,

    @OneToMany(mappedBy = "squad")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Squad(id=$id, name=$name, tribe=$tribe)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Squad
        if (id != other.id) return false
        return true
    }

    companion object {
        val UNASSIGNED_SQUAD = Squad(fromString("00000000-0000-0000-0000-000000000000"), "Unassigned Members")
    }
}
