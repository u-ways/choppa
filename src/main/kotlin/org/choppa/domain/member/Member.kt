package org.choppa.domain.member

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.chapter.Chapter
import org.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "member")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
