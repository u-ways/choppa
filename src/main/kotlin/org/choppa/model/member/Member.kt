package org.choppa.model.member

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.choppa.model.chapter.Chapter
import org.choppa.model.history.History
import org.choppa.model.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
data class Member @JsonCreator constructor(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("id")
    val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    @JsonProperty("name")
    val name: String,

    @ManyToOne(fetch = EAGER)
    @JsonProperty("chapter")
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter,

    @ManyToMany(mappedBy = "members", fetch = EAGER)
    @JsonIgnore
    val squads: MutableList<Squad> = mutableListOf(UNASSIGNED_SQUAD),

    @OneToMany(mappedBy = "member", fetch = LAZY)
    @JsonIgnore
    val iterations: MutableList<History> = mutableListOf()
) {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"
}
