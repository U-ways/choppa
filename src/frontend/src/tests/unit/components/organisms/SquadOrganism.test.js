import { describe, expect, it, shallowMount } from "@/tests/unit/test-imports";
import Squad from "@/data/types/squad";
import Member from "@/data/types/member";
import SquadOrganism from "@/_old/components/organisms/SquadOrganism";

const testSquad = new Squad(6, "Squad Two", "#00FFFF", [
  new Member(10, "Johnny"),
  new Member(11, "Robert"),
]);

describe("Squad Organism test", () => {
  it("Renders correctly", async () => {
    const wrapper = shallowMount(SquadOrganism, {
      propsData: {
        squad: testSquad,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
