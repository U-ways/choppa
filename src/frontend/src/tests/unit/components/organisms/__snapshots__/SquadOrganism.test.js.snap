// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Squad Organism test Renders correctly 1`] = `
"<div class=\\"squad\\">
  <div class=\\"squad__header\\" style=\\"border-top-color: #00FFFF;\\">
    <h1 class=\\"squad__header-name m-0\\"><span class=\\"squad__header-name--light\\">Squad</span> Squad Two</h1>
    <div class=\\"row align-items-center m-0 pt-1\\">
      <div class=\\"squad__header-member-count\\">
        2 Members
      </div>
      <div class=\\"squad__header-profile-pictures ml-1\\">
        <profilepicturestackmolecule-stub members=\\"[object Object],[object Object]\\" showcount=\\"3\\"></profilepicturestackmolecule-stub>
      </div>
    </div>
  </div>
  <div class=\\"squad__contents\\">
    <div class=\\"row ml-0 mr-0 pb-1\\">
      <div class=\\"col-1\\"></div>
      <div class=\\"col-6\\"><span class=\\"squad__contents--heading\\">Name</span></div>
      <div class=\\"col\\"><span class=\\"squad__contents--heading\\">Chapter</span></div>
    </div>
    <div class=\\"row squad__contents-person-row\\">
      <div class=\\"col-6 d-flex align-items-center squad__contents-person-cell\\">
        <div class=\\"d-inline mr-3 squad__contents-person-picture\\">
          <profilepictureatom-stub member=\\"[object Object]\\"></profilepictureatom-stub>
        </div>
        <div>Johnny</div>
      </div>
      <div class=\\"col-3 squad__contents-person-cell\\"></div>
    </div>
    <div class=\\"row squad__contents-person-row\\">
      <div class=\\"col-6 d-flex align-items-center squad__contents-person-cell\\">
        <div class=\\"d-inline mr-3 squad__contents-person-picture\\">
          <profilepictureatom-stub member=\\"[object Object]\\"></profilepictureatom-stub>
        </div>
        <div>Robert</div>
      </div>
      <div class=\\"col-3 squad__contents-person-cell\\"></div>
    </div>
  </div>
</div>"
`;
